Implement Plugin Classes

burzum opened the pull request:
* Plugins have a class that describes bootstrapping and routes
* Plugin objects are attached to a plugin registry in Application.php

This is still WIP but I opened the PR already to get some feedback before going on with this.

burzum committed:
Started working on a new plugin loading system

burzum committed:
Allow plugins to disable / enable bootstrap and routes

burzum committed:
Refining the PluginRegistry

burzum committed:
Refactored the way plugins are initialized

ADmad commented on src/Core/PluginRegistry.php:
```diff
@@ -0,0 +1,172 @@
+    public function console($commands)
+    {
+        foreach ($this as $plugin) {
+            if ($plugin instanceof ConsoleApplicationInterface) {
```
Since you have `PluginApp implements ConsoleApplicationInterface` doesn't this check become redundant? IMO the base class should not implement `ConsoleApplicationInterface`, only plugins which provide console commands should implement that interface.

burzum replied:
Correct @robertpustulka

ADmad replied:
Gotcha, my bad.

ADmad commented on src/Core/PluginApp.php:
```diff
@@ -0,0 +1,144 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function disableRoutes($disabled)
```
It would be better to have 2 separate methods `disableRoutes()` and `enableRoutes()`.

ADmad commented on src/Http/BaseApplication.php:
```diff
@@ -117,4 +168,14 @@ protected function getDispatcher()
+     *
+     * @return \Cake\Core\PluginRegistry
+     */
+    public function pluginRegistry()
```
`getPluginRegistry()`? :)

ADmad commented on tests/TestCase/Core/PluginRegistryTest.php:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    /**
+     * @expectedException \Error
```
This won't work on PHP 5.6 since it does not have an `Error` class.
https://travis-ci.org/cakephp/cakephp/jobs/319198618#L917

markstory commented on src/Core/PluginRegistry.php:
```diff
@@ -0,0 +1,172 @@
+    public function middleware($middleware)
+    {
+        foreach ($this as $plugin) {
+            $middleware->add($plugin);
```
Is the thinking here that plugins will *be* middleware? Or that plugins can *add* middleware. I think we want plugins to *add* middleware, which would look like

```php
foreach ($this as $plugin) {
  if ($plugin instanceof HttpApplicationInterface) {
    $middleware = $plugin->middleware($middleware);
  }
}
return $middleware
```

markstory commented on src/Core/PluginRegistry.php:
```diff
@@ -0,0 +1,172 @@
+    public function __invoke(ServerRequestInterface $request, ResponseInterface $response, $next)
+    {
+        foreach ($this as $plugin) {
+            $response = $plugin($request, $response, $next);
```
This implementation is at odds with the `middleware()` method. I'm not convinced that plugin classes should *be* middleware yet.

burzum replied:
OK, but the method must exists because of the interfaces.

markstory commented on src/Http/BaseApplication.php:
```diff
@@ -73,6 +120,8 @@ public function routes($routes)
             require $this->configDir . '/routes.php';
             // Prevent routes from being loaded again
+
+            $this->pluginRegistry()->routes($routes);
```
I would prefer to see us load routes in the application's routes file. This allows the app developer the ability to add additional paths/parameters via scopes. In 3.5 we added `RouteBuilder::loadPlugin()` to load routes from a single plugin. If the route builder had access to the plugin registry, or application instance we'd be able to leverage that existing interface.  With that said we may also need a `$routes->loadPlugins()` to cover the case where people just don't want to customize routes and just load them all.

burzum replied:
Any suggestion how to have a single plugin registry and share it between the router and the app without having a global DI container?

markstory replied:
We could hide the PluginRegistry inside `Plugin`. That would let us make the new/old ways to work with plugins easier to deal with.

> I understand your arguments but it don't think this functionality should exist in two places?

I don't think we should plan to have the same functionality in two places as a permanent solution. Doing so requires us to document when to use each solution and ensure that both of them continue working.

> Also it's already loading the routes if the plugin is configured to load them? I don't see the issue except that we would lose a previously added methods feature?

The routebuilder methods make it easier to customize plugin routes with additional scopes/paths. For example

```php
$routes->scope('/tools', function ($routes) {
  $routes->loadPlugin('DebugKit');
});
```
Is not possible with the current implementation you have. This approach prevents plugin routes to be placed above routes defined in the application's route list. Placing frequently used routes higher in the file helps improve parse performance as fewer routes are checked on incoming requests.

robertpustulka replied:
From the lack of a better designed solution making `Plugin` a “facade” could work for the time being. We can rethink that in the future (4.x/5.x) when we have a DIC built in.

burzum replied:
>  In 3.5 we added RouteBuilder::loadPlugin() to load routes from a single plugin. 

Can't we check in the method if the first passed was a plugin registry object and then use the 2nd args as string for the name?

> $routes->loadPlugins()

Could take the plugin registry as arg as well.

burzum committed:
Corrections to the plugin loading process

burzum committed:
Some stickler and phpcs fixes

thinkingmedia commented:
It's always great to see people working on new features, but PR's make it difficult to have a discussion. You can keep this PR up (absolutely nothing wrong with that), but if you want *feedback* it's better to create an issue and link it to this PR.

When you open the issue. Besure to explain how your changes make life easier for your fellow Cake developers, or explain what kind of problem you're solving.

markstory commented on src/Core/PluginInterface.php:
```diff
@@ -30,18 +30,30 @@ public function bootstrap();
+     *
+     * @return $this
+     */
+    public function enableRoutes();
```
I like these method names :clap:

burzum committed:
Minor fixes to Plugins

burzum committed:
Adding Plugin.loaded event

ADmad commented on src/Core/PluginRegistry.php:
```diff
@@ -0,0 +1,194 @@
+                'plugin' => $plugin,
+                'config' => $config
+            ]);
+            EventManager::instance()->dispatch($event);
```
Once #11692 is done the application event manager instance could be injected into plugin registry instead of fetching the global instance through static method.

burzum replied:
This might become a little off topic but by looking at this PR and the use case(s), I think we begin to use it like a DI container? Wouldn't it be better to wait for that?

markstory replied:
I think we can separate off events from this. #11692 will at solve several event manager related requirements.

markstory committed:
Convert PluginRegistry into a Collection.

A simpler collection will work for Plugins. This will let us integrate
new style plugins into `Plugin` more easily as we have a simple
container object to work with.

markstory committed:
Add names to plugins & remove middleware interfaces.

Plugins do not have to implement the Application interfaces as they are
hook based plugins and not the same as applications.

While they could be the same, it doesn't make sense for plugins to add
themselves as a middleware layer when they have a middleware hook. The
Application adds itself as a middleware to ensure there is an 'end' to
the middleware queue and ensure that a response is always generated.

markstory committed:
Integrate new plugin hooks into the console runner

Add the new plugin hooks to the console runner.

markstory committed:
Add interface for applications using plugins

Adding another interface will let applications not use plugins if they
don't want to, and also customize how their plugin hooks are implemented
if they really need to.

markstory committed:
Integrate plugin hooks in http & routing

Use the new hooks in the correct places to allow plugin routes,
bootstrap to be loaded after the application's configuration. This
allows plugins to define routes and configuration based on what the
application is doing/has set up.

markstory commented:
I've pushed up a bunch of changes that make this code more inline with what was documented in the related issue. I've not fully implemented the `addPlugin()` method yet, but will be doing that shortly.  What is 'done' is the plugin collection (instead of a registry) and the plugin hooks being called. Still to be done are:

* [ ] Actually loading plugins.
* [ ] Options and methods for the other plugin hooks
* [ ] Injecting the plugin collection in to `Plugin` for backwards compatibility
* [ ] Connecting the plugin collection into the route builder so that `RouteBuilder::loadPlugin()` works as it should.

markstory committed:
Implement addPlugin()

markstory committed:
Implement option passing for PluginApp

Using options is a simple way to integrate plugin settings with other
forms of configuration like config files.

markstory committed:
Add remaining hook methods to PluginApp and PluginInterface

robertpustulka commented on src/Core/PluginApp.php:
```diff
@@ -0,0 +1,236 @@
+     */
+    public function __construct(array $options = [])
+    {
+        foreach (['bootstrap', 'routes', 'console', 'middleware'] as $key) {
```
Can use [`VALID_HOOKS`](https://github.com/cakephp/cakephp/pull/11564/files#diff-56af3672217e52636f93d3273b3daa1fR38) const here

markstory replied:
Good idea. Fixed.

markstory committed:
Use constant.

markstory committed:
Add getPath() to plugins.

Plugins need to know their paths in order to be compatible with the
current features offered by Plugin.

markstory committed:
Make exceptions raised by PluginCollection better.

Raise MissingPlugin so we can give better error pages.

markstory committed:
Add class & config path.

These are also required by the current Plugin system. And are
requirements to get parity.

markstory committed:
Make plugin names an option.

This will make it possible to use PluginApp as an 'auto-plugin'
placeholder.

markstory committed:
Integrate new & old style plugins.

Old style plugin loading (via `Plugin`) now uses the new style plugin
classes. It won't use concrete plugin classes should they exist as old
style plugins are likely not updated.

This does allow plugins loaded via `BaseApplication::addPlugin()` to be
available to `Plugin` and thus work with locators/iterators using that
class.

markstory committed:
Fix use of undeclared property.

robertpustulka commented on src/Core/PluginInterface.php:
```diff
@@ -0,0 +1,141 @@
+     *
+     * @return void
+     */
+    public function bootstrap();
```
Should this interface define other hooks (`console` etc) as well? 

Or maybe it should be split into `ConsolePlugin` and `HttpPlugin` interfaces? I'm wondering if we should keep different interfaces for plugins and regular app. App is nicely  defined through interfaces (console, http, upcoming events) but here we just throw everything into one bucket.

markstory replied:
What benefits are there in having 3 interfaces for plugins? With the application interfaces it enabled standalone console applications to not concern themselves with http. This may also have been a side-effect of the `BaseApplication` being in the Http namespace. With plugins the base class is in Core along with the interface.

robertpustulka commented on src/Http/BaseApplication.php:
```diff
@@ -52,6 +68,63 @@ public function __construct($configDir)
+    public function pluginMiddleware($middleware)
+    {
+        foreach ($this->plugins->with('middleware') as $plugin) {
+            $middleware = $plugin->middleware($middleware);
```
This assumes that the `middleware()` method is implemented by the plugin app class, but the interface does not define it. This applies to other hooks as well.
Either we should add hook methods into the `PluginInterface` or check for methods existence inside `with()` method.

markstory replied:
middleware missing from the interface is a mistake. I'll fix that up.

robertpustulka commented on src/Core/PluginCollection.php:
```diff
@@ -0,0 +1,151 @@
+     */
+    protected $plugins = [];
+
+    const VALID_HOOKS = ['routes', 'bootstrap', 'console', 'middleware'];
```
Should we add another valid hook into the array once #11692 is merged?

markstory replied:
Yes

robertpustulka reviewed (commented):
I'm not sure we should mix http with console in one `PluginInterface`. This seems inconsistent with the design of current application interfaces.

robertpustulka commented:
Would it make sense to support any kind of hook by the plugin app? Instead of having concrete `enableRoutes()` etc methods we could have `enable($hook)`, `isEnabled($hook)` etc similar to the `with($hook)`. This way the interface wouldn’t mix the CLI and web context and allow plugins to hook into authentication for example.

markstory committed:
Add missing methods to PluginInterface and make it narrower

Define fewer methods to enable/disable hooks and instead use generic
methods. This does mean that autocompletion won't work for hook names,
but it means hooks could be extended more easily in the future. I also
prefer the narrower interface.

stickler-ci committed:
Fixing style errors.

robertpustulka commented:
Recent changes look good :+1: 

There are number of PHPStan reported issues though.

robertpustulka committed:
Fix issues reported by PHPStan

robertpustulka committed:
Fix more issues reported by PHPStan

markstory commented:
Thank you for fixing the phpstan errors @robertpustulka 👏

markstory committed:
Merge branch '3.next' into 3.next-plugins

markstory committed:
Add events hook to plugins.

This adds consistency between application hooks and plugin hooks.

markstory commented:
@robertpustulka I've added in an `events` hook for plugins now.

markstory commented:
> EventApplicationInterface returns the manager instance. Do we want to do the same here?

We could, but then callers would need to overwrite the manager on themselves and the application to ensure that state is consistent. Overwriting event managers also runs the risk of dropping listeners in the case a plugin accidentally creates a new manager. Should we remove the overwriting from the application events hook?

robertpustulka commented:
> Overwriting event managers also runs the risk of dropping listeners in the case a plugin accidentally creates a new manager.

The same can happen with routes or middleware as the hooks must return objects. I think that behavior of all hooks that modify an object should be similar.

robertpustulka commented on src/Http/BaseApplication.php:
```diff
@@ -61,6 +78,75 @@ public function __construct($configDir, EventManagerInterface $eventManager = nu
+    /**
+     * {@inheritDoc}
+     */
+    public function pluginEvents()
```
I think that this method should behave similar to the `pluginMiddleware` and `pluginRoutes` ie accept an object and return the modified one.

The risk of dropping changes if a new instance is returned by mistake is the same as with the middleware and routes.

dereuromark replied:
Immutable approach of always using return value to continue sounds indeed more consistent.

markstory replied:
Ok. I will get this updated. Do you also think all the new hooks should have typehints added where we have existing interfaces?

markstory replied:
I've added typehints for `EventManagerInterface` as it was the only existing interface.

robertpustulka replied:
@markstory `pluginEvents()` hasn't been updated with type hints.

robertpustulka replied:
> It also doesn't take parameters or return a value either though.

`pluginRoutes` and `pluginMiddleware` take an argument and return the modified object. I don't like that `pluginEvents` is different. Either we remove that from routes and middleware, or add to events.

> Perhaps the Server and CommandRunner should drop their event managers and instead proxy the application manager?
In this case the application would need to support events. This is optional now, as http and console application interfaces don't extend `EventDispatcherInterface`.

robertpustulka replied:
> It also doesn't take parameters or return a value either though.

`pluginRoutes` and `pluginMiddleware` take an argument and return the modified object. I don't like that `pluginEvents` is different. Either we remove that from routes and middleware, or add to events.

Also having `pluginEvents` with no arguments assumes that application that implements `PluginApplicationInterface` must have it's own event manager. Executing that method on an "eventless" application instance would take no effect. The idea of application "hooks" in my opinion is to take an object, modify it and return the modified instance.

> Perhaps the Server and CommandRunner should drop their event managers and instead proxy the application manager?

In this case the application would always need to support events. This is optional now, as http and console application interfaces don't extend `EventDispatcherInterface`.

robertpustulka replied:
I feel like I need to explain my point a bit more.

In "isolation" application does not dispatch events by default (it doesn't have to implement the `EventDispatcherInterface`), the Server and CommandRunner do. So in this scenario event hooks should be executed on Server's and CommandRunner's event manager. So both `events` and `pluginEvents` must take an argument in order to modify the server/runner's event manager. 
If application does implement `EventDispatcherInterface` it's event manager is injected into the server/runner in `setApp` setter. The process of applying hooks on an event manager is still the same, but the subject (event manager instance) is different now (app's event manager)

markstory replied:
I hear you on it feeling inconsistent, my concern is how much logic adding a parameter/return value forces onto the caller of `pluginEvents()`. For example the bootstrap method on both `Server` and `CommmandRunner` would need to look like:

```php
        $this->app->bootstrap();
        if ($this->app instanceof EventApplicationInterface) {
            $eventManager = $this->app->events($this->getEventManager());
            $this->setEventManager($eventManager);
        }
        if ($this->app instanceof PluginApplicationInterface) {
            $this->app->pluginBootstrap();
            if ($this->app instanceof EventApplicationInterface) {
                $events = $this->app->getEventManager();
                $events = $this->app->pluginEvents($events);

                $this->app->setEventManager($events);
                $this->setEventManager($events);
            }
        }
```

To me this puts a lot of onus on the caller and breaks the encapsulation that the interfaces should be trying to provide. Perhaps the `PluginApplicationInterface`should include/consume the `EventApplicationInterface` and `EventManagerInterface`? That would allow us to have fewer checks across the framework. It would prevent 'only' using events and not plugins, but I think that is a reasonable tradeoff to make for reduced complexity.

Joining the interfaces may also allow us to make proxying eventmanager methods simpler on `Server` and `CommandRunner`. The `getEventManager()` method could have an implementation that looks something like:

```php
public function getEventManager()
{
  if ($this->app instanceof PluginApplicationInterface) {
    return $this->app->getEventManager();
  }
  return EventManager::instance();
}
```

dereuromark commented on src/Core/PluginInterface.php:
```diff
@@ -0,0 +1,117 @@
+     * @param \Cake\Routing\RouteBuilder $routes The route builder to update.
+     * @return \Cake\Routing\RouteBuilder
+     */
+    public function routes($routes);
```
Could we use some interface of RouteBuilder here as typehint?

markstory replied:
An interface for RouteBuilder seems wasteful to me as it is not something we expect/want users to reimplement.

dereuromark replied:
OK

dereuromark commented on src/Core/PluginInterface.php:
```diff
@@ -0,0 +1,117 @@
+     * @param \Cake\Event\EventManager $events The application's event manager
+     * @return void
+     */
+    public function events($events);
```
Could we use some interface of EventManager here as typehint?

markstory committed:
Make plugins need to return the event manager

Plugin events() hook now works like the application one. This improves
consistency between plugins and applications.

markstory committed:
Add more tests around PluginApplication methods.

Fix mistakes as `__DIR__` is not going to work.

markstory commented:
What do folks think about the name `PluginApp`? Is `ApplicationPlugin` a better name?

What about the informal convention of having `src/Plugin.php` in each plugin. Should that instead be `src/Debugkit.php` where 'DebugKit' is the plugin's name?

rchavik commented:
@markstory what about namespaced plugin?

burzum commented:
@markstory if we move the whole plugin related classes into `Cake\Plugin`? We can have a `Plugin` class there as well then and deprecate the one in `Cake\Core`?

Let's use `ApplicationPlugin`. And I prefer to have a file `Plugin` in each plugin, makes it easy to remember and spot and also tells you what it is about.

saeideng commented:
>having src/Plugin.php in each plugin

markstory committed:
Rename PluginApp -> BasePlugin

This better mirrors the base class name for applications.

markstory committed:
Allow plugins to be loaded by short plugin names.

ADmad commented on src/Http/BaseApplication.php:
```diff
@@ -61,6 +78,79 @@ public function __construct($configDir, EventManagerInterface $eventManager = nu
+     */
+    public function makePlugin($name, array $config)
+    {
+        if (!class_exists($name)) {
```
Could just check if name does not have backlash instead of class_exists(), similar to what App::className() does.

ADmad commented on src/Http/BaseApplication.php:
```diff
@@ -61,6 +78,79 @@ public function __construct($configDir, EventManagerInterface $eventManager = nu
+        if (!class_exists($name)) {
+            $name = str_replace('/', '\\', $name) . '\\' . 'Plugin';
+        }
+        if (!class_exists($name)) {
```
Wouldn't PHP already generate a good enough error message if you try to instantiate a non existent class below?

markstory replied:
Could, I thought this error was better though.

markstory committed:
Simpler/faster short classname check.

markstory committed:
Merge branch '3.next' into 3.next-plugins

markstory committed:
Require files more than once.

Plugins will need their routes and setup run multiple times during
integration tests. With a require once this is not possible.

markstory committed:
Add missing file header.

markstory committed:
Merge the EventApplication and PluginApplication interfaces.

They had an awkward relationship that is better expressed by making them
the same. This also lets the pluginEvents() hook be a 'transparent' hook
method similar to routes, middleware and console.

markstory committed:
Make Server/CommandRunner proxy event managers.

Instead of copying state around, we can use the existing interfaces to
create proxies for evented applications. This lets us make evented and
non-evented applications work the same as far as the
Server/CommandRunner are concerned.

markstory commented:
@robertpustulka I've updated these changes to include the merged interface and proxies we talked about today.

markstory committed:
Add missing typehint.

markstory commented:
@robertpustulka I've updated these changes to include the merged interface and proxies we talked about today.

robertpustulka commented on src/Core/PluginApplicationInterface.php:
```diff
@@ -0,0 +1,80 @@
+     * @param \Cake\Event\EventManagerInterface $eventManager Event manager instance.
+     * @return \Cake\Event\EventManagerInterface
+     */
+    public function events(EventManagerInterface $eventManager);
```
Since now `PluginApplicationInterface` extends `EventDispatcherInterface` would it make sense to remove argument and return form `events ` and `pluginEvents` method? App that supports event hooks now has it's own event manager and listeners should be attached to that instance anyway.

```php
// App\Application.php
public function events()
{
    $this->getEventManager()->on(new MyListener);
}
```

and

```php
public function pluginEvents()
{
    $events = $this->getEventManager();
    foreach ($this->plugins->with('events') as $plugin) {
        $events = $plugin->events($events); //this stays as it is, plugin doesn't dispatch events
    }
    $this->setEventManager($events);
}
```

This way event handling could be even more encapsutaled.

markstory replied:
I'm ok with that. I added the argument to make the routes, events, and middleware hooks consistent in that they all get a 'thing' and return that 'thing'.  Making them 'closed' would add inconsistency with other hooks but simplify framework code.

robertpustulka replied:
Yeah, I know I’ve been advocating for consistency but now with the current API this makes little sense to have event hooks get and return something since it’s going to be always an app’s dependency.

markstory committed:
Remove events hooks.

Now that plugin applications support events out of the box, we can
remove the additional hook and leverage the bootstrap hook to attach
events. This makes applications and plugins simpler.

markstory committed:
Fix phpstan warnings and actual mistakes.
